### DISCUSSION TITLE
From TypeOrm to Prisma migration

### DISCUSSION THREAD

**Discussion by MaloLebrin** [sidebase/sidebase#43] [Ideas]
What do you think about migrating from typeOrm to prisma? Because typeorm is not well maintained, and has a lot of weak points on a large scal.

**Comment by BracketJohn** [comment]
What advantages do you see for prisma / what disadvantages do you see for TypeORM?

We've used both and working with prisma was super cumbersome because:
- the DB schema is not defined in TS but in their own customer language -> this means another learning steps for devs who want to work on your project!
- the type-generate step that is necessary to run every time you change your schema slows down the development process,
- some backends and some queries are not easy to support / write with it, e.g., json operations are very limited

This is why after first using prisma, we switchted to TypeORM. With ~30k stars and  the most recent commit being from 4 days ago, it also seems rather well maintained to us (:

**Reply by mahdiboomeri** [reply]
I haven't used typeORM in production but I can give you my perspective on Prisma
I think that [prisma docs](https://www.prisma.io/docs/concepts/more/comparisons/prisma-and-typeorm) make a really good argument about this.

- About the schema language, I believe that Prisma's DSL's learning curve is close to non-existent due to its simplicity (compare to type decorators) and it can take advantage of the VSCode extension which can be a huge boost in productivity (autocompletion and auto-generating relations)
- While the type-generation can be annoying in some situations, if it results in better type safety that's a trade-off I'm willing to make
- After using Prisma in a few production apps I haven't found any limitations in the query language I'd like to see you elaborate on this one further

**Reply by ruheni** [reply]
Hey 👋🏽

Alex here, from the Prisma team.

Adding on to what @mahdiboomeri responded with, Prisma has a gentle learning curve and is relatively easy to use.

> some backends and some queries are not easy to support/write with it, e.g., json operations are very limited

@BracketJohn, could you share some of the example backend/ queries you had challenges with? We recently promoted JSON filtering to General Availability a few weeks ago in v4.0.0.

**Reply by ruheni** [reply]
Also, let me know how it goes if you end up migrating to Prisma.

Feel free to [ping me](https://twitter.com/ruheni_alex) any time if you run into any issues. I'm also happy to jump on a call to have an in-depth discussion if needed, provide any clarifications, and learn about your getting started experience. 🙂

**Reply by BracketJohn** [reply]
Hey you two 👋 

Will ping you when it comes up! I think, putting my gripes with prisma aside, it would be great to allow people to use it when they want to (: So, let's integrate prisma as soon as we have enabled the option to choose setup, e.g., as part of introduction of `create-nuxt3-app`, which will be more of a "wizqard" style approach.

See #13 for the discussion on this

**Reply by HigherOrderLogic** [reply]
This is my two cents about TypeORM's disadvantage:
- The schema is defined using `Entity` class, and on connect, TypeORM automatically sync schema and database, which would cause unexpected result
- Poor performance on large scale
- Database randomly disconnect
- No ESM support

**Reply by BracketJohn** [reply]
> The schema is defined using Entity class, and on connect, TypeORM automatically sync schema and database, which would cause unexpected result

this is false / only true when misconfigured: you can turn this setting on / off using the `synchronize` option

> Poor performance on large scale
> Database randomly disconnect
> No ESM support
 
do you have evidence for these?

---

anyways, coming back to this, probably @ruheni: I found some internal gripes we had when using prisma in the past:
- [x] no support for rollbacks -> I think this was fixed and there is support now (see https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/generating-down-migrations)
- [ ] prisma does not support the programmatic actions: you absolutely have to use the CLI for this, you can't do `prisma.migrate('up')` -> unresolved, I think one can still only use the CLI?
- [ ] prisma does not allow reactive updated by assingment, e.g., the following pattern does not worL
    ```ts
    audit.status = 'resolved'
    prisma.commit()
    ```
    - this forces one to always write queries using the prisma client
- [ ] JSON fields are not supported well, e.g., postgres style select support still outstanding: https://github.com/prisma/prisma/issues/2431

Each individual one of these would probably not be a blocker for us, but the combination of them lead to a unpleasent experience when we tried to use Prisma in our project - maybe we're missing something that addresses these?

**Reply by djschilling** [reply]
i would like to add to @BracketJohn points.

We also started using Prisma in a project and later migrated to TypeOrm.
For us the auto generated schema did not work from time to time and debugging this was really a pain.
TypeOrm does less magic and i can fix most of the errors that appear there more easily.

So i like the decision to use TypeOrm

**Reply by HigherOrderLogic** [reply]
> do you have evidence for these?

For ESM support, it has been added in https://github.com/typeorm/typeorm/pull/8536. I haven't used TypoORM in a long time so my information was wrong. And about poor performance on large scale and randomly disconnect, this is something I saw when deploying a Discord bot with about 200 requests per sec, and I don't keep the log so I don't have any evidence to show.

**Reply by BracketJohn** [reply]
Welp - I've also had longer discussions with people reaching out on discord (https://discord.gg/9MUHR8WT9B) and we figured that it now is best to move to prisma ORM 🎊 

So, good news for most people who are part of this thread. Track #80 for progress on this.

**Reply by BracketJohn** [reply]
Switch is completed -> @MaloLebrin your initial ask should now be completed ❤️ 💯  Consider to mark this question as answered if there is nothing more.
